aeneas/src/mach/ShadowStackSpiller.v3

```diff
@@ -295,29 +295,29 @@ class ShadowStackSpiller() {
 		}
 		return false;
 	}
-	def succsHasKill(succs: Array<SsaCfEdge>) -> bool {
-		for (e in succs) {
-			if (hasKill(e.dest)) return true;
+	def killsFromSucc(block: SsaBlock, succs: Array<SsaCfEdge>) -> BitMatrixRow {
+		var killRow = BitMatrixRow(matrix, lpPassStart + block.mark);
+		killRow.clear();
+		for (i < lps.length) if (lps[i].0 == block) {
+			killRow.ori(order.order.length + lps[i].2.livepoint);
 		}
-		return false;
-	}
-	def hasKill(block: SsaBlock) -> bool {
-		var livepoints = codegen.livepoints;
-		for (i < livepoints.length) {
-			if (livepoints[i].0 == block) return true;
+		for (e in succs) {
+			killRow.ori(lpPassStart + e.dest.mark);
 		}
-		return false;
+		killRow.and(allLive);
+		return killRow;
 	}
 	def pushRestoreUp(info: SsaBlockInfo) {
 		var block = info.block;
 		curRow = BitMatrixRow(matrix, reloadStart + block.mark);
 		var succs = block.succs();
+		var killRow = killsFromSucc(block, succs);
 		var child = info.dom_child;
 		if (child != null) {
-			if (child.dom_sibling == null || hasKill(block) || succsHasKill(succs)) {
+			if (child.dom_sibling == null || killRow.count() > 0) {
```

Comment:
We can make this a bit more efficient if instead of `count()` we have `isZero()` for a row.